### PR TITLE
WIP: Jg refinement bot actions

### PR DESCRIPTION
**Pull request by jasongornall**
## Explanation

Explanation:

1. the Github action yaml is what triggeres the initial flow (will be catching labels and opened issues) Currently set as a workflow so I can test it. This action will either grab a random recent issue missing an estimate / needs a priority or catch new stuff coming in. [Currently only set to workflow](https://github.com/sourcegraph/sourcegraph/pull/35560/files#diff-ee9f5f3e7367c0ca4aad3622792b1bc23d8730ff7b816a1e31bb2cbd860f7678R6)

2. SendMessage is the core fire and forget flow that sends the message into a respective channel ([example](https://sourcegraph.slack.com/archives/C03BV5AKHL6/p1653330648606969)).

3. The Actions folder is responsible for responses from the slack refinement bot. This operates as a separate independent app with it's own deployment mechanism. [Selecting Priority](https://github.com/sourcegraph/sourcegraph/pull/35560/files#diff-31350f96d3e81dad84db00ed5f0cfdf0b541abfaf0d4b5b6e098919a94934f5dR75) and[ Selecting Estimate](https://github.com/sourcegraph/sourcegraph/pull/35560/files#diff-31350f96d3e81dad84db00ed5f0cfdf0b541abfaf0d4b5b6e098919a94934f5dR134) hit the respective callbacks

## Test Plan

I'd like to iterate a bit on this and try this out with a manual workflow.. if I'm happy phase 2 I'll put it on auto pilot. I'm currently proxying through my local computer and it seems to be working [well](https://sourcegraph.slack.com/archives/C03BV5AKHL6/p1653330648606969)

## App preview:

- [Web](https://sg-web-jg-refinement-bot-actions.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-zfrokuexvj.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: dev/refinement-bot/README.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Sourcegraph release tool
+
+This directory contains scripts and code to automate our releases. Refer to
```
Comment (jasongornall) [inline dev/refinement-bot/README.md:3]: Todo update this

File: dev/refinement-bot/actions/functions/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,215 @@
+import { App, ExpressReceiver } from "@slack/bolt";
+import * as functions from 'firebase-functions';
+import axios from "axios";
+
+// @ts-nocheck
+console.log('panda')
```
Comment (jasongornall) [inline dev/refinement-bot/actions/functions/src/index.ts:6]: obviously I'll get rid of this

File: .github/workflows/slack-refinement-notify.yml [hunk 3]
```diff
@@ -0,0 +1,89 @@
+
+name: Issue Labeled
+on:
+
+  # will turn this on for real once I test post merge
+  workflow_dispatch:
```
Comment (jasongornall) [inline .github/workflows/slack-refinement-notify.yml:6]: will enable this for now just so I can test it all together..

File: dev/refinement-bot/actions/.firebaserc [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
+  "projects": {
+    "default": "protean-iridium-350015"
```
Comment (jasongornall) [inline dev/refinement-bot/actions/.firebaserc:3]: node of product in firebase functions

**Comment by umpox** [comment]
@jasongornall Can you run `yarn format` in the root? It looks like some files haven't been prettier-ified 😆  Will make it easier to review. We should have `editor.formatOnSave` enabled by default in VSCode so not sure why that hasn't happened here 🤔 

Side note: I wonder if it makes sense to move this logic to a different repo? It isn't directly related to the Sourcegraph product, and we could use this in across multiple repos so might be better suited in its own dedicated repo.

**Comment by jasongornall** [comment]
> @jasongornall Can you run `yarn format` in the root? It looks like some files haven't been prettier-ified 😆 Will make it easier to review. We should have `editor.formatOnSave` enabled by default in VSCode so not sure why that hasn't happened here 🤔
> 
> Side note: I wonder if it makes sense to move this logic to a different repo? It isn't directly related to the Sourcegraph product, and we could use this in across multiple repos so might be better suited in its own dedicated repo.

The reason I can't is the github action. I COULD add the action itself to another repo but then I have to split the bot inhalf :/

File: dev/refinement-bot/actions/functions/src/index.ts (hunk 2 above)
Comment (taylorsperry) [reply dev/refinement-bot/actions/functions/src/index.ts]: 🐼

**Comment by umpox** [comment]
@jasongornall Ah, does the action have to live in this repo in order to pick up new issues/listen to labels? Makes sense to keep it here if so, I thought maybe this could live above all of our repos and listen to everything - I think 99% of our issues live here though :)

File: dev/refinement-bot/actions/functions/package.json [hunk 5]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "ngrok-dummy",
```
Comment (umpox) [inline dev/refinement-bot/actions/functions/package.json:2]: We have a lot of `package.json` files here (3 within `/refinement-bot` folder).

Generally we:
1. Declare package.json dependencies in the [root package.json](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@jg_refinement_bot_actions/-/blob/package.json?L360%3A4-360%3A20=) unless we need to publish something (let the build handle the rest)
2. Declare scripts in relative `package.json` fiels (e.g. web/package.json)
3. Name the relative ones in the format `@sourcegraph/refinement-bot`

My suggestions would be (if possible):
- Only have 1 package.json for this files (e.g. in refinement-bot/package.json). Delete the other ones and name the root one appropriately. Add all scripts to this file.
- Move the dependencies to the root file

File: dev/refinement-bot/actions/functions/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    priority: string
+}
+export interface EstimateValue {
+    issue: string
+    estimate: string
+}
+
+// @ts-nocheck
+// const config = functions.config();
+require('dotenv').config()
```
Comment (umpox) [inline dev/refinement-bot/actions/functions/src/index.ts:21]: I don't think we should need `dotenv` if we're just accessing `process.ENV` and not using a `.env` file 🤔

File: dev/refinement-bot/actions/functions/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+})
+
+app.action('button', async ({ ack, say, action }) => {
+    await ack()
+    await say('Yaaay! that command works!')
+    // Update the message to reflect the action
+})
+
+const updateUrlIfChanged = async (issueNumber: string, channelId: string, messageTs: string) => {
+    const issue = await axios.get(`https://api.github.com/repos/sourcegraph/sourcegraph/issues/${issueNumber}`, {
```
Comment (umpox) [inline dev/refinement-bot/actions/functions/src/index.ts:51]: We're using both `axios` and `node-fetch` in these files, can we change to just use `node-fetch`? Should carry over without any issues afaik, we don't have `axios` as an explicit dependency

File: dev/refinement-bot/actions/functions/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+        headers: {
+            Authorization: `token ${process.env.REFINEMENT_BOT}`,
```
Comment (umpox) [inline dev/refinement-bot/actions/functions/src/index.ts:53]: Do we pass in `REFINEMENT_BOT` anywhere? We should probably rename to `REFINEMENT_BOT_SECRET` :D

File: dev/refinement-bot/actions/functions/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+        }
+    )
+}
+
+app.action('priority_select', async ({ ack, say, action: actionBase, respond, body: bodyBase }) => {
+    await ack()
+
+    // types seem to be off here...
+    const body: any = bodyBase
+    const action: any = actionBase
```
Comment (umpox) [inline dev/refinement-bot/actions/functions/src/index.ts:82]: I think adding the dependencies to the root package.json will help fix this (assuming they had type `any` and the types aren't just actually incorrect from the library)

File: dev/refinement-bot/actions/functions/yarn.lock [hunk 10]
```diff
@@ -0,0 +1,2711 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (umpox) [inline dev/refinement-bot/actions/functions/yarn.lock:1]: We should only really need one `yarn.lock` in the root of the repository, will be solved when shifting dependencies up and we can delete this file

File: dev/refinement-bot/src/getRandomIssue.ts [hunk 11]
```diff
@@ -0,0 +1,62 @@
+import { setOutput, setFailed } from '@actions/core'
+import { getOctokit } from '@actions/github'
+import { Repository } from '@octokit/graphql-schema'
+import type { Endpoints } from '@octokit/types'
+import dotenv from 'dotenv'
+dotenv.config()
```
Comment (umpox) [inline dev/refinement-bot/src/getRandomIssue.ts:6]: Shouldn't need this lib unless we require a `.env` file, (can still get `process.env` without this)

File: dev/refinement-bot/src/getRandomIssue.ts [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        if (!estimates) {
+            throw new Error('missing estimates')
+        }
+    } catch {
+        setFailed('unable to get estimate labels')
+        return
+    }
+
+    // step 2 get issues without estimates that are on ${team}
+    const query = `repo:"sourcegraph/sourcegraph" is:issue is:open label:${team} -label:${estimates} sort:updated-desc no:assignee -label:tracking`
```
Comment (umpox) [inline dev/refinement-bot/src/getRandomIssue.ts:41]: There's a possible bug where `estimates` could contain undefined which would result in stuff like: ",test,,", I think?

I like this version below. The `isDefined` helper is just something which tells `.map` that `item.name` definitely exists now. We use it in a few places

```suggestion
    const isDefined = <T>(value: T): value is NonNullable<T> => value !== undefined && value !== null
    const estimates = result.repository.labels?.nodes?.filter(isDefined).map(item => item.name) || []

    if (estimates.length === 0) {
        setFailed('unable to get estimate labels')
        return
    }
    // step 2 get issues without estimates that are on ${team}
    const query = `repo:"sourcegraph/sourcegraph" is:issue is:open label:${team} -label:${estimates.join(
        ','
    )} sort:updated-desc no:assignee -label:tracking`
```

File: dev/refinement-bot/src/sendMessage.ts [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+    process.exit()
+}
+
+const isPriorityPresent = (json_issue: GithubIssue): boolean => {
+    let priorityPresent = false
+    priorityPresent = !!json_issue.labels?.some(label =>
+        // match p0, p1, p2
+        priorityList.some((item: PriorityItem) => item.value === label?.name)
+    )
+    return priorityPresent
```
Comment (umpox) [inline dev/refinement-bot/src/sendMessage.ts:29]:
```suggestion
    if (!json_issue.labels) {
        return false
    }

    return json_issue.labels.some(label =>
        // match p0, p1, p2
        priorityList.some((item: PriorityItem) => item.value === label.name)
    )
```

Good chance where we can just leverage TS so we don't need `!!` to ensure the variable is not undefined

File: dev/refinement-bot/src/sendMessage.ts [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+const isEstimatePresent = (json_issue: GithubIssue): boolean => {
+    let estimatePresent = false
+    estimatePresent = !!json_issue.labels?.some(label => label.name.startsWith('estimate/'))
+    return estimatePresent
```
Comment (umpox) [inline dev/refinement-bot/src/sendMessage.ts:35]: Same as above here, checking that `json_issue.labels` isn't undefined first will clean this up a lot!

File: dev/refinement-bot/src/yarn.lock [hunk 15]
```diff
@@ -0,0 +1,4 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (umpox) [inline dev/refinement-bot/src/yarn.lock:1]: Should be able to delete this file

File: dev/refinement-bot/yarn.lock [hunk 16]
```diff
@@ -0,0 +1,8 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (umpox) [inline dev/refinement-bot/yarn.lock:1]: Should be able to delete this file

**Review by umpox — COMMENTED** [review]
Did a first pass, really interesting concept and looking forward to seeing it in action, have left some comments :)

After looking further, I do think this would be better suited in a different repo. I feel like having this on a cron is probably enough to meet our needs right now. Issues generally come quite slowly and don't change often so the likelihood of it getting out of sync might actually be quite rare. WDYT? It'd mean we could use this on other repos too which might be valuable (thinking of the customer repo in particular here!). Left some related comments in `slack-refinement-notify.yml`

File: dev/refinement-bot/actions/functions/src/index.ts (hunk 6 above)
Comment (jasongornall) [reply dev/refinement-bot/actions/functions/src/index.ts:21]: this just made it easier for me to test locally admittedly haha

File: dev/refinement-bot/actions/functions/yarn.lock (hunk 10 above)
Comment (jasongornall) [reply dev/refinement-bot/actions/functions/yarn.lock:1]: so actions are an entirely seperate deployment with seperate dependencies. Think of it as an independent app from scratch

**Pull request closed without merging**